orc/src/main/java/com/netflix/iceberg/orc/TypeConversion.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2018 Hortonworks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.iceberg.orc;
+
+import com.netflix.iceberg.Schema;
+import com.netflix.iceberg.types.Type;
+import com.netflix.iceberg.types.Types;
+import org.apache.orc.TypeDescription;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TypeConversion {
+
+  /**
+   * Convert a given Iceberg schema to ORC.
+   * @param schema the Iceberg schema to convert
+   * @param columnIds an output with the column ids
+   * @return the ORC schema
+   */
+  public static TypeDescription toOrc(Schema schema,
+                                      List<Integer> columnIds) {
```

Comment:
Is there a better way to pass column IDs back? Isn't this dependent on the order you use to traverse the Schema? Why not use a `BiMap` between column ID and full column names?

In other places, we use a `BiMap`. One key part of that is that you only ever convert in one direction: from components like `["a", "b"]` to a full name, `"a.b"`. That way, you never have to parse the names and avoid the problem of fields named "a.b".

Reply:
Ugh. I avoid Guava at all costs. They are *really* bad at breaking API compatibility.

I haven't done the schema evolution stuff yet, so maybe. I mostly need these so that I can stringify it to store it into the file.

All ORC TypeDescriptions have automatically assigned ids that run 0 (for the root) to N-1 for the right most leaf. So mapping from the ORC to Iceberg id is easy given the list.

Reply:
My main concern is that this depends on the order of traversal, which is easy to accidentally break. Since this is a critical piece of information, I'd like to see a different solution to track IDs. It doesn't have to be guava (though we do use guava elsewhere because it provides so much value).

Reply:
The relevant constraint is that the columnIds.get(orcSchema.getId()) gives you Iceberg id.

This code does depend on the fact that ORC's ids are assigned sequentially given a prefix ordering. On the other hand, changing that would be a serious change to ORC's API.

But fine, I'll change this to a Map.